### ISSUE TITLE
Activations Exploding Across Layers

### ISSUE THREAD

**Issue by c3-utsavdutta98** [allenai/OLMo#797]
### ❓ The question

I was curious if there are any explicit mechanisms in place to prevent activations norms exploding with the initialization olmo2 uses.
Specifically, with a N(0,0.02) init, followed by `x = x + norm(attn(x, ...))` layers, I would assume the variance of activations should keep increasing over layers? The paper suggests otherwise though in Section 3.2.

Unfortunately, when I initialize a random model, I don't see the same behaviour and observe constant increase in activation variance.

Was wondering if someone from the team could shed light on what prevents this from happening?

**Comment by dirkgr** [comment]
Section 3.2 does not graph activations, it graphs gradients, and it shows the gradient norm across training steps, not through the layers. We have not checked whether activations grow as you go through the layers, but they should. Every layer should increase activation norm by `2*sqrt(2)`. 

There is an unresolvable tension here. If you want the activations (on the residual stream) to always stay in the same window, how can you add to it from MLPs and attention? You can scale the contribution and the residual stream, but then you diminish the contributions of the earlier layers. In practice, it seems that the growth by `sqrt(2)` per MLP/attn is fine, but there could be something better out there.

**Comment by c3-utsavdutta98** [comment]
Thanks for the reply!
And yeah, it seems with this norm layout the growth is inevitable.

I wonder if you tried re-norming on nthe main branch (this was proposed in the swin norm paper, where they do it every 6 layers).

Also, with reference to 3.2; this was the excerpt I was referring to : 

<img width="650" alt="Image" src="https://github.com/user-attachments/assets/aff51990-243b-4d1c-b4ab-79bb64c1b404" />

Where it says the "growth_exponent" stays near 0 for **both** activations and gradients? Which would mean, I assume, that the activations are also not scaling up in norm (somehow?).

**Comment by loftusa** [comment]
Here's a pretty basic check on this. I got the activations in every layer for a single prompt, then averaged over batch and hidden dimension to get the average mean and variance per layer.

Mean:

<img width="459" alt="Image" src="https://github.com/user-attachments/assets/63978875-2781-4ca8-90d7-9d3d7227b855" />

Variance:
<img width="454" alt="Image" src="https://github.com/user-attachments/assets/c82e5302-78e7-458a-935c-81b4f67e7783" />

Code (I used the `[nnsight](https://www.nnsight.net/)` library):
```
#%%
from transformers import AutoModelForCausalLM, AutoTokenizer, AutoConfig
import nnsight
from nnsight import LanguageModel
import torch

model_name = "allenai/OLMo-2-1124-7B"
lm = LanguageModel(model_name)
#%%
prompt = "The capital of France is"
with lm.trace(prompt):
    activations = nnsight.list().save()
    for l in lm.model.layers:
        activations.extend(l.output)

activations = torch.stack(activations).squeeze()
#%%

import numpy as np
import matplotlib.pyplot as plt
variances = activations.var(dim=(-1, -2)).numpy(force=True)
means = activations.mean(dim=(-1, -2)).numpy(force=True)
#%%
plt.plot(means, marker="o")  # Add circular markers at each data point
plt.xlabel("layer")
plt.ylabel("mean of activations, \naveraged over batch and hidden dimension")
plt.grid(True)  # Add grid lines for better readability
plt.show()
#%%
plt.plot(variances, marker='o')  # Add circular markers at each data point
plt.xlabel("layer")
plt.ylabel("variance of activations, \naveraged over batch and hidden dimension")
plt.grid(True)  # Add grid lines for better readability
plt.show()
#%%
```

**Comment by dirkgr** [comment]
Why does the mean of the activations start below zero?

**Comment by loftusa** [comment]
> Why does the mean of the activations start below zero?

Why wouldn't they?

**Comment by dirkgr** [comment]
I thought you'd be plotting the mean of the _absolute_ values. Otherwise it doesn't really show the effect you're after, which is (I think) that the magnitude of the activations grows from one layer to the next. You get that with your norm and variance graphs, but I'm not sure the _mean_ graph contributes to my understanding of what's going on.

Anyways, sorry for the delay, but I find this super interesting. Can you make the same graphs for the 13B and 32B? I'm asking because your variance graph looks really scary, and it would lead you to believe that it's impossible to scale much deeper. But the 13B and 32B are much deeper models and they worked fine. I'm curious to see what happens to the activations there.
